docs-devsite/database.datasnapshot.md

```diff
@@ -38,7 +38,7 @@ export declare class DataSnapshot
 |  [child(path)](./database.datasnapshot.md#datasnapshotchild) |  | Gets another <code>DataSnapshot</code> for the location at the specified relative path.<!-- -->Passing a relative path to the <code>child()</code> method of a DataSnapshot returns another <code>DataSnapshot</code> for the location at the specified relative path. The relative path can either be a simple child name (for example, "ada") or a deeper, slash-separated path (for example, "ada/name/first"). If the child location has no data, an empty <code>DataSnapshot</code> (that is, a <code>DataSnapshot</code> whose value is <code>null</code>) is returned. |
 |  [exists()](./database.datasnapshot.md#datasnapshotexists) |  | Returns true if this <code>DataSnapshot</code> contains any data. It is slightly more efficient than using <code>snapshot.val() !== null</code>. |
 |  [exportVal()](./database.datasnapshot.md#datasnapshotexportval) |  | Exports the entire contents of the DataSnapshot as a JavaScript object.<!-- -->The <code>exportVal()</code> method is similar to <code>val()</code>, except priority information is included (if available), making it suitable for backing up your data. |
-|  [forEach(action)](./database.datasnapshot.md#datasnapshotforeach) |  | Enumerates the top-level children in the <code>IteratedDataSnapshot</code>.<!-- -->Because of the way JavaScript objects work, the ordering of data in the JavaScript object returned by <code>val()</code> is not guaranteed to match the ordering on the server nor the ordering of <code>onChildAdded()</code> events. That is where <code>forEach()</code> comes in handy. It guarantees the children of a <code>DataSnapshot</code> will be iterated in their query order.<!-- -->If no explicit <code>orderBy*()</code> method is used, results are returned ordered by key (unless priorities are used, in which case, results are returned by priority). |
+|  [forEach(action)](./database.datasnapshot.md#datasnapshotforeach) |  | Enumerates the top-level children in the <code>IteratedDataSnapshot</code>.<!-- -->Because of the way JavaScript objects work, the ordering of data in the JavaScript object returned by <code>val()</code> is neither guaranteed to match the ordering on the server nor the ordering of <code>onChildAdded()</code> events. That is where <code>forEach()</code> comes in handy. It guarantees the children of a <code>DataSnapshot</code> will be iterated in their query order.<!-- -->If no explicit <code>orderBy*()</code> method is used, results are returned ordered by key (unless priorities are used, in which case, results are returned by priority). |
```

Comment:
I think the placement of this is off; "neither" should be placed right before the first instance of "the ordering" in order to clarify which things we are contrasting. As it stands, it would be more logical to say something like "neither guaranteed nor recommended," which is not what we mean.

Here and below.

Reply:
I don't think your suggestion makes sense. I'll drop the change.

Reply:
I dunno Josh, ever since I scored in the 99th percentile of the verbal GRE with a 780, I've been pretty confident about writing the English -- and that was a long time ago!  :D

To clarify, I'm not suggesting "neither guaranteed nor recommended" -- that's just an example of how to resolve the tension created by "neither guaranteed."

What I *would* suggest, is one of these:

a) "not guaranteed to match . . ."

b) "is guaranteed to match neither the ordering on the server, nor the ordering of `onChildAdded()` events."

And I absolutely DO guarantee that both of those are grammatically correct.  I'd vote for a) any day due to its economy and simplicity, but you seemed to be showing a liking for "neither" constructions.

Reply:
Sorry, the problem with your initial suggestion is that there were _three_ `the ordering`s in the text, and I suspect you actually wanted to point me to the _second_ one.

Here's my proposal to reduce the complexity there (I'm not include the markup or the other sentence as it doesn't matter -- I'd have to apply any changes to multiple places if they're accepted anyway):

> JavaScript objects do not guarantee that the order of data returned by val() will match either order of events from the server or from onChildAdded() events.

I'm happy to restore some words at the beginning if I'm wrong in my guess that those other words I've dropped aren't particularly essential.

That said, if you don't think that changes to this text will be fruitful, then I'd rather leave the changes dropped (as they are right now) and focus on the remainder of the PR.

Reply:
Probably leaving it as/is works best.  The suggested text feels like is lacking a definite article before "order of events," so I'd stay away from that phrasing.